### ISSUE TITLE
sticky: extensions

### ISSUE THREAD

**Issue by Thorin-Oakenpants** [ghacksuserjs/ghacks-user.js#492]
previous threads #294 #211 #12 
woo... the old issue of 294 is a palindrome of this issue 492 ... spooky :ghost: 

---

Use this issue for extension announcements: new, gone-to-sh*t, recommendations for adding or dropping in the wiki list [4.1: Extensions](https://github.com/ghacksuserjs/ghacks-user.js/wiki/4.1-Extensions). Stick to privacy and security related items

### :small_orange_diamond: possible additions

* [Modify Header Value](https://addons.mozilla.org/firefox/addon/modify-header-value/)
* [mHeaderControl](https://addons.mozilla.org/firefox/addon/mheadercontrol/) | [GitHub](https://github.com/maheee/mHeaderControl)
* [Referer Control](https://addons.mozilla.org/firefox/addon/referercontrol/)
* [History AutoDelete](https://addons.mozilla.org/firefox/addon/history-autodelete)
* [WebAPI Manager](https://addons.mozilla.org/firefox/addon/webapi-manager/) | [GitHub](https://github.com/snyderp/web-api-manager)
* [Forget Me Not](https://addons.mozilla.org/firefox/addon/forget_me_not/) | [GitHub](https://github.com/Lusito/forget-me-not)
* [CookieBro](https://addons.mozilla.org/firefox/addon/cookiebro/)
* [Luminous](https://addons.mozilla.org/firefox/addon/luminous/) | [GitHub](https://github.com/gbaptista/luminous) - JS Events handling
* FoxyProxy
   * [Basic](https://addons.mozilla.org/firefox/addon/foxyproxy-basic)
   * [Standard](https://addons.mozilla.org/firefox/addon/foxyproxy-standard/)

### :small_orange_diamond: nah <sup>feel free to discuss</sup>

* [Canvas Defender](https://addons.mozilla.org/firefox/addon/no-canvas-fingerprinting)
   * nah. see https://github.com/ghacksuserjs/ghacks-user.js/issues/458 , Canvas Blocker is far superior
* [Google search link fix](https://addons.mozilla.org/firefox/addon/google-search-link-fix/) | [GitHub](https://github.com/palant/searchlinkfix)
   * google/yandex. Personally, I suggest using something more universal
* [VTzilla](https://addons.mozilla.org/en-US/firefox/addon/vtzilla/)
   * shitty privacy policy, you have to opt-out of scanning everything, but could be handy for some people, as long as they are aware of the privacy implications
* [Searchonymous2](https://addons.mozilla.org/firefox/addon/searchonymous2/) | [GitHub](https://github.com/DaAwesomeP/Searchonymous2)
  * use containers if you want to separate google search from logged in google services
* [eCleaner (Forget Button)](https://addons.mozilla.org/firefox/addon/ecleaner-forget-button/) | [GitHub](https://github.com/andy-portmen/ecleaner/)
   * abandoned since Sept 2017, and outstanding issues. Forget Me Not looks better for this kind of thing
* [Shape Shifter](https://addons.mozilla.org/firefox/addon/shape-shifter/) | [GitHub](https://github.com/snapper26/shapeshifter)
   * abandoned since Aug 2017, doesn't really offer anything not already covered
* [Negotiator](https://addons.mozilla.org/firefox/addon/negotiator/) | [GitHub](https://github.com/hultqvist/negotiator)
  * abandoned since July 2017 (and it had some issues/flaws according to the big E)

...

**Comment by Thorin-Oakenpants** [comment]
I'll add an `extensions [tools]` section to the wiki and we can add items like cloudflare detect. And I'll go back thru the old sticky and amend OP with any info

**Comment by claustromaniac** [comment]
I think another decent candidate for a `[tools]` section would be [uBO-Scope](https://github.com/gorhill/uBO-Scope).

Well, if the idea is to separate privacy/security-related but non-protecting extension into a separate list, then extensions like the legacy SSleuth would belong in there too, right? BTW, I kinda miss SSleuth.

**Comment by Thorin-Oakenpants** [comment]
> the idea is to separate privacy/security-related but 

There's already a legacy [tools] section. The idea that tools don't actually make changes. SSleuth is in there. When it becomes a working Web Ext, I can move it out of legacy. I have the repo in my stars, and there is slow progress to get the apis

**Comment by Just-me-ghacks** [comment]
3P Request Blocker - Page not found

**Comment by practik** [comment]
re https://github.com/ghacksuserjs/ghacks-user.js/issues/294#issuecomment-388491739

> Forget Me Not …  work very unreliable

@crssi , do you remember what problems you found with FMN?  I ask because I've been trying it out for about a week (after using Cookie AutoDelete for nearly a year), and so far FMN has been just as good as CAD, in some ways even better.  What should I be watching out for?

**Comment by Atavic** [comment]
@Just-me-ghacks https://addons.mozilla.org/en-GB/firefox/addon/tprb/

**Comment by crssi** [comment]
@practik 
Yes. What I have found back then is that FMN worked well until it decided to not to.
It was very random, sometimes few minutes after browsing, sometimes an hour or so, but at one moment it stopped to delete cookies after, until FF restart.
Since I have not returned to FMN now for months, it might be already sorted out.
I would say for you to just check, every once and a while, if cookies are removed as they should.
Cheers

**Comment by practik** [comment]
^^ !! That's bad. But it hasn't done that for me so far. Hopefully it is sorted out, it's gone through a few updates since you tested it. I'll keep an eye on it. Thanks!

**Comment by grauenwolfe** [comment]
Whoa, [Luminous](https://addons.mozilla.org/en-US/firefox/addon/luminous/) looks like it could be badass. Anyone using it already?

**Comment by Atavic** [comment]
I have it on an install used for unlogged browsing (like no github or webmail) only.

I block events like `beforeunload` and all the events related to mouse movements, like `mouseover`, etc.

See: https://gbaptista.github.io/luminous/doc/en-US/

**Comment by grauenwolfe** [comment]
@Atavic 
Very down-to-earth but still entirely thorough documentation. Hopefully it can actually help me deal with all of the internet's rampant annoyances. Probably try it out soon on another machine and see how it does.

**Comment by Thorin-Oakenpants** [comment]
> BTW, I kinda miss SSleuth.

- `https://github.com/sibiantony/ssleuth/issues/78#issuecomment-392271507`
- example extension: https://addons.mozilla.org/en-US/firefox/addon/certainly-something/
- API ticket: https://bugzilla.mozilla.org/show_bug.cgi?id=1322748

So no idea when or what SSleuth Web Ext will look like

**Comment by claustromaniac** [comment]
Ah, great. I was about to mention that I renamed my repo and now the link to Detect Cloudflare PA should be broken, but it seems Github is smart enough to redirect folks to the new URL. :tada: 

I still want to mention that I went ahead and [listed it on AMO](https://addons.mozilla.org/firefox/addon/detect-cloudflare-plus/). Traktofon seems to be MIA or something, and I was bored, so I also added a toolbar icon to it and made the address bar icon optional, among many other thingies. 

So far it works great for me. The only significant issue left to fix seems to be that it can't always behave as expected when the backward or forward navigation actions are used, but that one seems kinda painful to fix compared to the other issues that I already fixed. I may eventually work on that, though.

Anyway, I thought you may want to know.

**Comment by Thorin-Oakenpants** [comment]
^^ your github readme needs to link to your AMO, preferably at the top and before linking to Detect Cloudflare

**Comment by claustromaniac** [comment]
> its not really a fork anymore IMO

What would you call it?

> ^^ your github readme needs to link to your AMO, preferably at the top and before linking to Detect Cloudflare

There _was_ already a link at the top, but not in the readme. People seem to ignore those, though...

> "This extension neither collects nor shares any kind of information whatsoever." <- needs a privacy header?

Added :heavy_check_mark:

**Comment by Kraxys** [comment]
I would like to advice for 3 extensions:

1) The 2 natural companions of I think every meticulous proxy or vpn user:
* **Change Timezone**:
https://addons.mozilla.org/en-GB/firefox/addon/change-timezone-time-shift
* **Location Guard**:
https://addons.mozilla.org/en-GB/firefox/addon/location-guard/

With these 2 addons you can make your system time zone and and wifi geolocation be in accordance with the IP geolocation and local time zone of the proxy/vpn server you are using. Not a one click process, though. But this avoids increasing your entropy by wearing a Russian ip and in the  the same time a system wifi geolocation and date settings that show you near Melbourne.

2) **BP Block Font Fingerprint**:
https://addons.mozilla.org/en-US/firefox/addon/bp-block-font-fingerprint
This extension avoid the detection of any font or any unique glyph, without impairing la appearance of the page (as setting the pref **browser.display.use_document_fonts** to **0** does).

**Comment by crssi** [comment]
0 font detection will make you unique for sure.

**Comment by earthlng** [comment]
re: BP Block Font Fingerprint:

> 0 font detection will make you unique for sure.

not only will it make you pretty unique because very few people use something like this extension,
the extension itself also has several flaws.
1. it comes with a absurdly long list of domains it considers "trusted" and which it allows to detect fonts if they want to. Among these domains are pretty much all google domains, twitter, facebook to name but a few ?!?!
2. it's easily detectable if sites cared to look. The functions it uses to overwrite certain things can be read
3. it's easily blockable with a CSP that doesn't allow inline JS in which case it's completely useless

but thanks to this extension suggestion I looked at fonts again in general and I found some things which I think we need to improve in the user.js. I'll open a new issue to discuss them

**Comment by Kraxys** [comment]
@crsi & earthing
I agree concerning the unicity created by BP Block Font Fingerprint. And thanks in particular to Earthing to make clear this extension had some fundamentals flaws.

What's your opinion concerning my suggestions about Change **TimeZone** and **Location Guard**, in order precisely to diminish the entropy raised by using a vpn/proxy server having time and location characteristics than those of the system's user?

For location, blocking geo wifi in preference must be considered as sufficient (except if the browsed site mandatory will geo wifi data, a case where the use of **Location Guard** could be useful). But not geoblocking and spoofing geo wifi with Location Guard according to the  proxy server used, set the user on a safer side in the point of view a spoofing: The location provided by the proxy server IP is in that case 
confirmed by geo wifi data sent by the browser, so reinforcing the its likehood.

Concerning **Change Timezone**, this extension solves a sharper problem as there isn't in FF's preference anything as "_don't send any date  time-zone data_" (as it was the  case for location in blocking geo wifi preference). Blocking these data from being sent could maybe be achieved with some uMatrix or NoScript setting, but it then raise an unicity flag, as not letting the browser send them  is not a common behavior.

**Comment by Atavic** [comment]
https://github.com/dessant/clear-browsing-data seems deep in options. Do they cover something interesting?

**Comment by Atavic** [comment]
https://github.com/dessant/clear-browsing-data seems rich in options.
Do they cover something interesting?

**Comment by Atavic** [comment]
https://github.com/dessant/clear-browsing-data seems rich in options.
Do they cover anything interesting?

**Comment by Kraxys** [comment]
**Clear Browsing Data** seems interesting. But after installing it, it seems not able, neither to clear browsing data when the browser closes, nor when it starts. Only during the browsing session.

In order to sanitize a browsing session as soon as it begins, there is [**StorageErazor**](https://addons.mozilla.org/en-US/firefox/addon/storagerazor): It clears Cache, Local Storage and IndexedDB each time the browser starts. The IndexedDB clearing is important, since 1) blocking IndexedDB in FF preference breaks some site 2) Cookies Autodeleted doesn't handle IndexedDB.

Maybe **Clear Browsing** Data and **StorageErazor** may be seen as complementary each other.

**Comment by Kraxys** [comment]
@practik :Thanks for this information. I didn't know checking "clear Offline Website Data" erased indexedDB. This strongly reduces the usefulness of StorageErazor, but I will nevertheless keep this addon enable **and** "clear Offline Website Data" checked, as the second works when the browser closes, and the first, when it starts, so that I'm absolutely sure to begin each browsing session on a neat basis :)

Other addons I suggest are the ones permitting to **block Authentication**:
Along  [ip-check.info](http://ip-check.info/description.php)
>"Many browsers allow web sites to send hidden authentication data to third party sites.  Example:
><link rel="stylesheet" type="text/css" href="http://Session:638431048@ipcheck.info/auth.css.php">

>This may either happen directly on the current page or in an iframe, and does NOT need JavaScript. If additionally iFrames and JavaScript are used, even the currently loaded page may get your ID. This data is deleted when the browser is closed, but, execpt for this, has the same effect as third party cookies.

>Your browser should not send any HTTP authentication data to third party sites.

>Currently known to be affected are: Chrome, Safari, Firefox".

I don't know whether or not all that is completely up-to-date, but if it remains true, I think it would be wise to prevent tracking via Authorization. 

I currently have found 2 addons permitting that:
[**Authentication Tracking Blocker**](https://addons.mozilla.org/en-US/firefox/addon/authentication-tracker-blocker) and [**Block Http Authentication**](https://addons.mozilla.org/en-US/firefox/addon/block-http-authentication)

Notice that blocking Authentication is one of the feature of [**Chameleon**](https://addons.mozilla.org/en-US/firefox/addon/chameleon-ext), too. This addon has many other interesting feature (as optionnally  spoofing time, screen size and ClientRects), and while using it **may** increase entropy, I think thant when **properly** used, it can in fact reduce it (eg when spoofing your system time accordingly the time of proxy server you are using, or when spoofing screen size with the most common ones for desktop PC, such as 1366x768 or 1920x1080).

**Comment by crssi** [comment]
@Kraxys do you have any example site using **Authentication**?

**Comment by Atavic** [comment]
[Here](https://github.com/swagger-api/swagger-ui/issues/686) the headers are described.

**Comment by crssi** [comment]
@Atavic  thank you, but I didn't mean a description but a real case site using it. :smile:

**Comment by Atavic** [comment]
I haven't seen any, you got to use Fiddler, Charles Proxy or similar tools to debug headers responses.

**Comment by crssi** [comment]
Using Fiddler here for years (now you made me to look at Charles Proxy, for which I have never heard before :smile:) and also found one at `https://www.amainhobbies.com/` over XHR. Interesting at this site is also that login doesn't work when EvilCorp analytics is blocked.. WTF.

**Comment by Atavic** [comment]
Charles is not free and has a Mac version. Privoxy is another proxy that changes or crunches headers.

**Comment by sanjayen** [comment]
Please change the Decentraleyes rules to add to uBlock Origin URL to https://git.synz.io/Synzvato/decentraleyes/wikis/Frequently-Asked-Questions#for-umatrix-and-ublock-origin-non-easy-mode-users

**Comment by Thorin-Oakenpants** [comment]
@earthlng .. FYI: `https://github.com/FirefoxBar/HeaderEditor/issues/45` resolved, for you to check out/test or whatever

**Comment by claustromaniac** [comment]
[Block Cloudflare MITM Attack](https://addons.mozilla.org/en-US/firefox/addon/bcma/) seems to have reincarnated about a month ago, by a different author. 

This one works differently than the previous one. I wonder for how long it will last this time.

**Comment by Thorin-Oakenpants** [comment]
OT: Interesting [read](https://www.htbridge.com/blog/FT500-application-security.html)

**Comment by Atavic** [comment]
^ There's an SSL Scan giving the full Certificate's chain: https://www.htbridge.com/ssl/
...and the Web Scan seems good, too.

**Comment by StanGets** [comment]
Hi,

I'm concerned about font fingerprinting and i'd appreciate your advice about a Firefox extension called [BP Block Font Fingerprint](https://addons.mozilla.org/en-US/firefox/addon/bp-block-font-fingerprint/).

I do have:
```
// 1401: disable websites choosing fonts (0=block, 1=allow)
user_pref("browser.display.use_document_fonts", 0);
```
nevertheless,
[Browserleaks' Font Fingerprinting](https://browserleaks.com/fonts) shows _Fingerprint for JS Fonts_ (unicode) and _Fingerprint for JS Fonts (classic)_

even if,
[JonDonym IP check](http://ip-check.info/?lang=en) shows _Fonts - 4 installed fonts have been found on your computer. - good_

When using the above mentioned BP Block Font Fingerprint no Fonts appear nowhere.
I'd appreciate your opinion on this extension. Thanks.

**Comment by Thorin-Oakenpants** [comment]
I checked this out recently, and it just breaks web pages - it seems to disable JS or something. I couldn't actually get the browserleaks page to work with the extension enabled. And I didn't bother to dig any deeper because at the end of the day I think this makes you SUPER unique (54 users)

https://addons.mozilla.org/en-US/firefox/addon/bp-block-font-fingerprint/reviews/
> Breaks JavaScript entirely - sad

`sad`, sounds like Trump :grimacing: . Anyway, I'm also concerned by this, and Client Rects, as high entropy items that need addressing. The only way to win against these two, is via numbers in RFP, IMO.

**Comment by StanGets** [comment]
> I think this makes you SUPER unique (54 users)
yeah, I had that in mind too, this entropy you often mention here. Seems to me this only dissuades from using the extension should it be harmless on the js scale.

I read the review about this extension mentioning it broke JavaScript entirely (sad! sounds trumpy when in fact it's basically a popular wording!) but I didn't understand exactly what that could mean once I had tried the extension and observed no broken javascript... I also had a look at the extensions script itself and noticed exceptions were handled for so-called trusted sites including google, apple ... that disturbed me a bit.

OK, got your advice, clear. Thanks @Thorin-Oakenpants

**Comment by Thorin-Oakenpants** [comment]
I'm not 100% on the uniqueness. I don't think it gives everyone's glyphs the same value. The question would be does this make you less unique than without? But TBH, without being able to get the test to work and the JS comment, I decided it wasn't worth looking at anymore - and I'm leery of too many extensions causing conflicts (eg. CSP).

**Comment by earthlng** [comment]
@StanGets see my comment [here](https://github.com/ghacksuserjs/ghacks-user.js/issues/492#issuecomment-423820620)

There are also still problems/limitations with WebExtensions that effectively allow sites to do things without extensions being able to do anything about it

**Comment by StanGets** [comment]
BP Block Font Fingerprint had already been evoked and discussed on this very thread. I wrote too quickly. Comments and yours in particular, @earthlng confirm that at this time the extension is not worth being installed. I had tested it only, quickly before uninstalling it and coming here because I was wondering if I was right to ... remove it.

Fast, quick, speedy... always running, should have checked. Thanks @Thorin-Oakenpants  and @earthlng

**Comment by garywill** [comment]
That BP Block Font Fingerprint seems not FOSS

> License
>     All Rights Reserved

I can't find source code link

**Comment by earthlng** [comment]
you can just download the addon, extract it and inspect the source code

**Comment by Thorin-Oakenpants** [comment]
^^ and sheeshus h christ .. just like everyone else starting out, they get basics wrong (leaking navigator objects, look at the closed issues) .. and the whole UA randomizing thing is such a turn off. Even chameleon is constantly patching "holes" with this. Sure it takes time for a product to mature. But why rely on one person when you're had years and many people looking at this already - Tor Uplift!

**Comment by atomGit** [comment]
no disrespect to the dev or what he's trying to do, but i'm glad you said that - one less thing to worry about

and yes, i know much of what is covered in Trace is covered by other necessary add-ons, but i don't know what all the dev is gonna do with this - i had it installed but with all the lists and some other stuff disabled

**Comment by Daystar1998** [comment]
Also, wouldn't Temporary Containers in auto mode plus FPI prevent most of the issues with keeping cookies, cache, IDB, etc. thus minimizing the need for an auto-deleter such as CAD? Especially when combined with using a whitelist for cookies. Or am I misunderstanding something with how containers and FPI work?

**Comment by crssi** [comment]
@Thorin-Oakenpants 

> control your JS people!

That is almost easy for you... geeks, but not for ordinary users. :smile_cat: 

@Thorin-Oakenpants and @Daystar1998 

From my observation (I am using FPI+TC auto mode)... and still there are some cookies left behind and CAD deals with those. It might be that a new TC doesn't pick those leftovers, but I haven't test that... yet.

**Comment by Thorin-Oakenpants** [comment]
@Daystar1998 Yes. TC can clean by a contextId API thingie which means IDB is cleared. I'll see if I can dig up it up .. ahhh https://github.com/ghacksuserjs/ghacks-user.js/issues/395 (in first post)

> Temporary Containers only uses one API to remove data, and that is [`contextualIdentities.remove`](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/contextualIdentities/remove) - which removes all [`userContextId` tagged storage (including IDB)](https://wiki.mozilla.org/Security/Contextual_Identity_Project/Containers#What_is_.28and_isn.27t.29_separated_between_Containers). 

And as stocially says, i a new container id = a new clean cookie/persistent data etc.

**Comment by Thorin-Oakenpants** [comment]
@crssi  "and still there are some cookies left behind and CAD deals with those"

Maybe you should look at why those cookies remain. If those cookies are no threat due to always starting a new container, then there's no need for CAD, just clear them on close. Seriously, cookie extensions cause way more problems than they could ever solve.

**Comment by crssi** [comment]
@Thorin-Oakenpants 
I can swear that there was a problem in the past, but now I haven't found any orphaned cookies... doesn't matter, I will drop the CAD for now.
And you are right (like doooh :smile: )... with TC, there is no need for CAD. I am in a bit of fear for TC future existence... it seems that @stoically activity dropped to zero. :crying_cat_face:

**Comment by Woofy-Wolf** [comment]
Kudos to @claustromaniac for two extensions that should IMHO be added to the wiki:

- [ETag Stoppa](https://addons.mozilla.org/en-US/firefox/addon/etag-stoppa/)
- [Privacy-Oriented Origin Policy](https://addons.mozilla.org/en-US/firefox/addon/privacy-oriented-origin-policy/)

Privacy-Oriented Origin Policy is damn impressive. ETag Stoppa and POOP replaced Header Editor and Cors Everywhere in my browser.

You already know the extensions, of course. I'm nominating them, here, to nudge them toward an updated wiki. :)

**Comment by Thorin-Oakenpants** [comment]
ETag Stoppa is already there. And the POOPer needs more work .. I'm waiting for the sticky issue to pass 200 comments as a sure sign.

@claustromaniac : Add the POOPer when you're ready

**Comment by crssi** [comment]
@Woofy-Wolf same here :smile:, the HE+Cors Everywhere combination were PoC/workaround until @claustromaniac  didn't made damn good PooP :smile_cat: 

@Thorin-Oakenpants IMHO PooP needs just a week or two of observation, but I think its ready.
For 100% mark, just a blacklist is missing, but its not a showstopper (and nothing is 100% in a real world).
From observation in the last couple of months, there is only `maxcdn.bootstrapcdn.com` to be added to blacklist for now... similar as it is now `www.youtube.com *.googlevideo.com` for whitelisting <- and nothing wrong with that, since its the same EvilCorp behind.
Now, with PooP and Referer control (or ghacks or over SR), I really ask myself what would be additional reason for 3rd-party font blockage from tracking/FP perspective? :wink: 

Cheers

**Comment by crssi** [comment]
@Thorin-Oakenpants 

> Now, with PooP and Referer control (or ghacks or over SR), I really ask myself what would be additional reason for 3rd-party font blockage from tracking/FP perspective?

Don't get me wrong... I didn't imply or call for a change in ghacks-user.js. It should stay as it is there IMHO.

**Comment by Woofy-Wolf** [comment]
> From my observation (I am using FPI+TC auto mode)... and still there are some cookies left behind and CAD deals with those.

This is my observation as well, and it's why I don't re-use containers in TC. I just tested a few sites. Cookies from washingtonpost.com were destroyed with their container. Cookies from nytimes.com and mozilla.com appeared to survive (which makes no sense whatsoever).

I used CAD until recently to delete cookies (& localstorage browser-wide) before their containers were destroyed. To delete cookies from a container that is about to be destroyed is (or should be) unnecessary and excessive. Anyway, I removed CAD after considering @Thorin-Oakenpants 's opinion of cookie managers, but I still take precautions because I don't understand what's going on. Instead of CAD, I set cookies to expire when the browser closes (minus some exceptions, each of which have their own permanent container). I run CCleaner, too. It always sees site data that FF didn't delete.

It's very strange.

**Comment by Thorin-Oakenpants** [comment]
> ccleaner.... It always sees site data that FF didn't delete

FYI: it lists HSTS data from SiteSecurityServiceState.txt as "cookies"

Some domains are restricted - i.e extensions cannot control them. Not sure if that included cookies
`extensions.webextensions.restrictedDomains default value` ( with my own line breaks added)
```
accounts-static.cdn.mozilla.net,
accounts.firefox.com,
addons.cdn.mozilla.net,
addons.mozilla.org,
api.accounts.firefox.com,
content.cdn.mozilla.net,
discovery.addons.mozilla.org,
input.mozilla.org,
install.mozilla.org,
oauth.accounts.firefox.com,
profile.accounts.firefox.com,
support.mozilla.org,
sync.services.mozilla.com,
testpilot.firefox.com
```

**Comment by atomGit** [comment]
speaking of ETags - the wiki says...

> Note: if you disable disk and memory cache, this is not required

should Temporary Containers be added to that?

not understanding the workings of all this stuff, i wanted to dbbl chk, so with TC set to auto and RFP disabled i tested (http://lucb1e.com/rp/cookielesscookies/) and came up clean

**Comment by crssi** [comment]
^^ First revisit shows last entered value, but page reload after that looks like clean.
Config here: TC(auto)+FPI+Memory enabled+Disk disabled
IDK but it looks like the first revisit still gets read from memory.

**Comment by atomGit** [comment]
interesting - wonder what the diff is?

TC=auto, FPI disabled, mem enabled, disk disabled

i don't see how that should make a difference and, besides, etags are stored in web cache, so how is it possible to read mem cache after a restart? did you restart, or just refresh?

**Comment by crssi** [comment]
Reopened in a new TC (the old destroyed).
Restart would be pointless for test when disk disabled.
But what I wrote on my previous post suggests something smelly with containers... its should be totally cleared (memory cache also separated in container) or totally not (memory cache not separated), but not something in between. FF bug maybe?

**Comment by claustromaniac** [comment]
> @claustromaniac : Add the POOPer when you're ready

It's no longer in *alpha*, which means I deem it complete and functional already. The documentation can use some more work, though.

Anyway, this may sound stupid to you, but I prefer not to add it to the wiki myself because it would be a biased recommendation. I don't have a problem recommending True Sight and ETag Stoppa, because the former merely provides information (it doesn't change the browser's behavior in any way), and the latter is extremely simplistic. However, POOP is fairly complex **and** it changes the behavior of cross-origin requests; it attempts to provide more privacy by altering a standard security mechanism.

I did a decent amount of research and I have confidence in my skills, but ideally a more experienced (read *actual*) developer should vet my work for me to feel comfortable recommending it myself. I don't expect that to happen, though. That would be quite a luxury.

> Now, with PooP and Referer control (or ghacks or over SR), I really ask myself what would be additional reason for 3rd-party font blockage from tracking/FP perspective? :wink:

You forgot to mention Decentraleyes! It doesn't cover all fonts (that would be insane) but I noticed it takes care of some of the very popular ones.

**Comment by Thorin-Oakenpants** [comment]
I'm actually afraid to use the POOPer ... it's installed and disabled. And I've ignored the long thread you guys chat in. I'm afraid of it breaking other extensions (rolling the dice thing) and it looks a little complex. 

I've been taking my dumps out in the garden, but I cover them up like a good :cat2: does

**Comment by claustromaniac** [comment]
I can give you a super long explanation of why I consider it safe, but I won't for the same reason I refuse to recommend it (even though I'm using it myself).

**Comment by Thorin-Oakenpants** [comment]
up to you :) now get over to the *other* project (and don't tell anyone) and help out

**Comment by atomGit** [comment]
Decentraleyes - web fonts is on the ToDo list... for like 2 years - he's running into some sort of problemo with that

> now get over to the other project (and don't tell anyone)

librefox - i'm liking the idea - haven't run it yet

**Comment by Thorin-Oakenpants** [comment]
No .. I have nothing to do with librafox

**Comment by claustromaniac** [comment]
I have nothing to do with LibreFox. (dafuq is libr**a**fox? :trollface:)

**Comment by atomGit** [comment]
> Anyway, this may sound stupid to you, but I prefer not to add it to the wiki myself because it would be a biased recommendation. 

wadda pussy (wee, i made a PUN!)

and you guys know damn well what Librefox is - i seen sum yas over there trying to jump ship!

so this whole add-on landscape changes with [TC](https://addons.mozilla.org/en-US/firefox/addon/temporary-containers/) - pants, i ass-u-me ur using this by now?

no more storage cleaners, but several other things removed as well

for the record, me=:

### MUST haves...

* [ClearURLs](https://gitlab.com/KevinRoebert/ClearUrls) - check this out verses Pure URL or whatever link cleaner you're using
* CSS Exfil Protection
* Decentraleyes, although the need for this is largely reduced by TC me thinks
* Firefox Multi-Account Containers - needed if you want to keep logins/settings/storage per-domain - supplements (a shortcoming?) in TC
* Skip Redirect
* Temporary Containers (currently running in auto-mode with FPI disabled)
* uBO
* uM - i REALLY want to dump this and use uBO exclusively - TC makes this (much?) more feasible - have to think some

### recommends...

* [ReFont](https://addons.mozilla.org/en-US/firefox/addon/refont/) - very slick - i haven't tested yet if web font fetching is blocked with this, but it seems it is
* Smart HTTPS (pants hates it - that's why i like it ... actually, i don't give shits if my webber hammers an http site because they should ALL be using https (and Let's Encrypt certs are free) so i let 'em know in the logs + it doesn't depend on a possibly inaccurate list + ...)

### not needed...

* ETag remover (TC)
* POOP (TC) ??? dunno about this - i asked on his repo - sure would hate to give up that little pile of pink poop on my toolbar : (
* window.opener be gone (TC) ?
* rel=noopener scripts/add-on (TC)
* Referer Control (TC, can relax pref in user.js - essentially can control referrer with TC)
* all/nearly all add-ons that have anything to do with auto-cleaning storage (TC)
* ...

lots of problems solved and other add-ons rendered obsolete with TC - so much so that it seems like this should be a requirement

**Comment by claustromaniac** [comment]
> and you guys know damn well what Librefox is

We know, we just aren't involved.

> POOP (TC) ??? dunno about this - i asked on his repo - sure would hate to give up that little pile of pink poop on my toolbar : (

[Answered](https://github.com/claustromaniac/poop/issues/18), in case anyone else wonders.

> - window.opener be gone (TC) ?
> - rel=noopener scripts/add-on (TC)

I think using those two at the same time is redundant... (gotta check)

**Comment by practik** [comment]
@atomGit, that's impressive how many things are made redundant by TC. I tested TC briefly but ditched it because it breaks Recently Closed Tabs, which I use all the time. But I may have to revisit that choice …

For those (few?) who are still interested in extensions for cleaning cookies etc.:  I've been using [Forget Me Not](https://addons.mozilla.org/en-US/firefox/addon/forget_me_not/) daily for the past four months and haven't encountered any of the [problems](https://github.com/ghacksuserjs/ghacks-user.js/issues/492#issuecomment-416054340) @crssi reported previously. And FMN does a couple of things better than Cookie AutoDelete:
* It allows you to whitelist cookies [by name](https://github.com/Cookie-AutoDelete/Cookie-AutoDelete/issues/68)
* It handles containers [more intelligently](https://github.com/Cookie-AutoDelete/Cookie-AutoDelete/issues/262)
* It clears [cookieless localStorage](https://mdn.github.io/dom-examples/web-storage/)

On the other hand, there's one thing CAD definitely does better: It clears cookies that are [set during redirects](https://github.com/Lusito/forget-me-not/issues/114). But even so I'd still say that FMN is just as Wiki-worthy as CAD.

That said, all this cookie monitoring I've been doing has been a big pain. Pants's [suggestion from last week](https://github.com/ghacksuserjs/ghacks-user.js/issues/492#issuecomment-449598392) –

> block all cookies and whitelist using FF's internals

– is looking extremely appealing.

**Comment by atomGit** [comment]
practik:

> tested TC briefly but ditched it because it breaks Recently Closed Tabs

i never use that, but your right (sort of - i see 2 entries and neither is correct) - history still works and you can access that via a toolbar button

there's niggles with TC - nav'ing back fails unless 'back' is in the same container

my biggest gripe is when you WANT to store web data (settings/cookies/log-ons) because you gota go through a SEVEN HUNDRED (5) step process - disable setting in TC>isolation>MAC; create a perm container; open site in perm; enable to always open site in perm; re-enable setting in TC - same for cross-domain log-ons i'm sure, at least some times, but you can pile a bunch of trusted sites in a single perm container too, like search engines, etc., then do all normal browsing in temp containers

for me the perks outweigh the pissers so far - i set uM to allow 1st party cookies globally, so that's the end of pissing around with cookies/storage/IDB/abc and def, then you can dump several add-ons and break shit less often - the ghacks user.js needs a few settings relaxed to make a smoothie but that, again, equals less breakage

**Comment by atomGit** [comment]
hey, i'm just gonna drop this here and let one of you pros handle it - probably mentioned long ago, but i couldn't quickly find anything and i didn't want to make a new issue...

nobody should be editing the user.js, right? so maybe have the updater script set read-only attrib on it after writing just as a reminder those who might forget?

**Comment by atomGit** [comment]
@practik said...
> I tested TC briefly but ditched it because it breaks Recently Closed Tabs ...

here's a potential half-fix: try setting 'Delete no longer needed Temporary Containers' to something other than 'after the last tab in it closes' - looks like after the time expires you still have to turn to the browser history though

**Comment by Thorin-Oakenpants** [comment]
https://github.com/ghacksuserjs/ghacks-user.js/issues/492#issuecomment-450505680
> Decentraleyes, although the need for this is largely reduced by TC me thinks

Heck no. You might want to edit that comment. DC is not about isolation, it's about reducing/eliminating 3rd party calls

**Comment by atomGit** [comment]
did that + moved POOP to required based on feedback from :cat:

**Comment by Thorin-Oakenpants** [comment]
https://github.com/ghacksuserjs/ghacks-user.js/issues/492#issuecomment-449598392
> I'm actually thinking of dropping CanvasBlocker myself. ClientRecs() is the only thing it adds for me

Actually, I CB is actually blocking a lot of canvas stuff that RFP isn't

**Comment by crssi** [comment]
@atomGit TC is really a good WE, but you cant "kill" all the thingy's with it.
I am a bit time limited, so will be my post. ;)

I don't like **ClearURLs**, since there is no user control and same, but with user control, can be achieved with **Neat URL** and **Skip Redirect**.
**Smart HTTPS** does NOT work with TC... at least not in automatic mode. Same goes with **HTTPS by default**. I don't like **HTTPS Everywhere**, really don't see a point to leak to EFF and most domains I use are even not listed in their database.
I don't see a point for permanent containers, so it is for **Multi-Account Containers**, I am using TC + **Context Plus**.
**Canvas Blocker** is more than just a Canvas blocker (also Audio and other), cannot be replaced with TC.
Not sure yet, if TC can replace **ETag remover**... I would say no.
**Referer Control** cannot be replaced by TC.

I really don't see a reason to disable FPI. In the past, there were plenty of problems, but lately Mozilla has really improved it and I could not find any breakages with FPI enabled.

You already got an answer for PooP and Decentaleyes.

Also with PooP, I do not see any reason (tracking wise) to block fonts and to complicate life with **Refont** or similar.

@practik with correct usage of TC, you can dump CAD and FMN and similar.

If you set, for example:
``` js
/* 1020  */ user_pref("browser.sessionstore.max_tabs_undo", 10);
```
you can undo closed tabs with CTRL+SHIFT+T and it will even reopen in the correct TC container (if invoked in a time before destroying the container).

**Comment by practik** [comment]
> if invoked in time before destroying the container

This was exactly my problem; there were too many times when I wanted to reopen a tab whose container was already gone. I think if I do try TC again I'd rather just retrain myself to stop using Recently Closed Tabs altogether.

**Comment by crssi** [comment]
@practik try something like this (starting at default settings)
```
    General
      Automatic Mode: Check
      Container Name Prefix: %domain% #
      Random Container Color: Check
      Container Icon: fingerprint
      Container Number: Reuse available numbers
      Delete no longer needed Temporary Containers: 2 minutes after the last tab in it closes
      Click [Save]
    Isolation
      Per Domain
        Domain pattern: *
          Always open in new Temporary Containers: enabled
          Click [Add or Edit]
        Domain pattern: duckduckgo.com
          Always open in new Temporary Containers: enabled
          Click [Add or Edit]
      Global
        Mouse Clicks on Links should open new Temporary Containers
          Middle Mouse: If the clicked Link Domain does not match the active Tabs Domain (Subdomains won't get isolated)
        Click [Save]
    NOTE: CTRL+SHIFT+T = Undo closed tab(s)
```

^^ I am using DDG for search.
Left mouse click will not open in a new container, but middle mouse click will if different domain.
This way you have a control to open in a new container and the same time cross domain logins will not fail with a mouse left click.
If you need to migrate container, then **Context Plus** extension will do.

Cheers

**Comment by crssi** [comment]
Nice one from @intika, making background of URL bar red color when non-secure site.
[Librefox HTTP Watcher - Red flag](https://addons.mozilla.org/firefox/addon/librefox-http-watcher-red-flag/)

**Comment by StanGets** [comment]
I don't know about you but here I always prefer a style over a script over an extension, when feasible of course.

Coloring the urlbar background given the site's security status is easily performed right from the user's userChrome.css file thanks to a few lines, available i.e. with [Color your URL bar based on website security](https://www.reddit.com/r/FirefoxCSS/comments/7dkhuh/color_your_url_bar_based_on_website_security/)

I chose (old habits I guess) to colorize one secure sites (yellow) perhaps because my memories go back to the times when secure sites were the minority, which is changing, fortunately.

**Comment by atomGit** [comment]
in this particular case i'm not sure that CSS accurately reflects the connection status, but yes, i was thinking the same thing - i recall reading something about something some time ago that suggested using CSS for this is a bad thing

**Comment by StanGets** [comment]
That invalidates what I thought was always true, that is that CSS is less likely to be harmful than jscript and jscript than an extension's potential.

Regarding colorizing the urlbar background given the site's security status I've never encountered issues but maybe because I refer only to `#identity-box.verifiedDomain` and `#identity-box.verifiedIdentity`.

I'm not a specialist as I spend my time repeating it (especially to those who know more, lol) and I learn every day.

**Comment by claustromaniac** [comment]
> **Smart HTTPS** does NOT work with TC... at least not in automatic mode. Same goes with **HTTPS by default**. I don't like **HTTPS Everywhere**, really don't see a point to leak to EFF and most domains I use are even not listed in their database.

@crssi, are you using any other extensions for that? Some time ago I made an extension that works somewhat similar to Smart HTTPS, but it is more minimalistic and gets along with TC. I'm still not sure I want to publish it (I hardly have any time for maintaining the stuff that I already published), but if there is any interest I'll consider it. I used HTTPS Everywhere for the longest time but it is heavy AF and I have come across numerous sites that weren't in its database...

**Comment by atomGit** [comment]
interest no.1 right here

**Comment by atomGit** [comment]
ditto - i should have also mentioned that - i wonder why in hates Moz hasn't done something about this long ago, standards and net etiquette be damned - this is a serious privacy issue for anyone not encrypting their traffic

**Comment by claustromaniac** [comment]
[There you go](https://addons.mozilla.org/en-US/firefox/addon/httpz/) ([source](https://github.com/claustromaniac/httpz)). It's ugly AF and it's not configurable, but it gets the shit done.

**Comment by Thorin-Oakenpants** [comment]
So I'm doing a dozen metric shit-tonnes of testing, and do you think I can find a decent working RFP toggle button? Yup, I am aware that some RFP effects require a restart.

Anyone know of one? Just a simple button that toggles RFP, nothing more.

**Comment by claustromaniac** [comment]
I don't know of any, but such an extension should be super trivial to make... If you don't find one by tomorrow or so, I'll look into it. :laughing:

**Comment by claustromaniac** [comment]
:jeans:, I know you didn't ask for this, but I literally made it while I was doing other totally unrelated stuff. It was *that* trivial. Change the file extension to XPI and install manually.

[rfp_switch-0.1.0-fx.zip](https://github.com/ghacksuserjs/ghacks-user.js/files/2730306/rfp_switch-0.1.0-fx.zip)

**Comment by Thorin-Oakenpants** [comment]
PS: reading this: https://medium.com/privateid-blog/privacy-inequality-the-most-brutal-form-of-inequality-youve-ever-imagined-e674d4f3cd42

I'm gonna be one of `The Protected`, not one of `The Predictables` - but it will still hurt: i am sure that a lack of data points will be held against people (getting a loan, insurance rates etc). It may pay to start curating a public ID of all things nice, sugar and spice....

Welcome to you new dark dystopian future...

**Comment by crssi** [comment]
@Thorin-Oakenpants 

> Just a simple button that toggles RFP, nothing more.

Just out of curiosity... Do you need this for testing purposes only (as a shortcut) or there are some real world usage?

**Comment by crssi** [comment]
@claustromaniac thank you for HTTPZ. :+1: 
I have one or two addons in mind, but really don't have a heart to ask, since I know you are busy with other stuff.

**Comment by Thorin-Oakenpants** [comment]
@crssi As I already said, it's useless IMO in a real world live main FF setup because it's a global switch and users have multiple tabs, multiple background js, and would have to check it's state when opening new web content - imagine making one mistake and leaving it off while you visited `murderers.4.hire` and googled `how to dissolved bodies`.  It just makes it easier for me testing in my test bed (different portable FF nilla profiles) `1` as a visual indicator and `2` as a super quick way to toggle it.

**Comment by Thorin-Oakenpants** [comment]
^^ Thanks StanGets .. I didn't bother thinking down the rabbit hole too deeply, because it's already a silly idea .. but that is just *another* reason a flip switching button to be used *in session* is a stupid idea

Lets say a user has RFP on, and uses the user.js with section 4600 commented out (as instructed). They then flip RFP off (manually or with an extension) to make one website work, and they would potentially give up more privacy because 4600 items are commented out. Yup, that's just stupid. And if they didn't have section 4600 commented out with RFP, then their fingerprint is F'd anyway.

But I am only using it in a test environment. With a vanilla profile, and no user.js. In fact in these tests I am only testing RFP

For the third time (3rd times a charm?) - this extension (and building it into other extensions as an option which I see a lot of) is a F stupid idea

**Comment by atomGit** [comment]
> PS: reading this: https://medium.com/privateid-blog/privacy-inequality-the-most-brutal-form-of-inequality-youve-ever-imagined-e674d4f3cd42
> 
> I'm gonna be one of `The Protected`, not one of `The Predictables` - but it will still hurt: i am sure that a lack of data points will be held against people (getting a loan, insurance rates etc). It may pay to start curating a public ID of all things nice, sugar and spice....

[SCS](https://en.wikipedia.org/wiki/Social_Credit_System) - the Chinese Social Credit System - coming soon to a country near you! - actually it's already here - many have been canned at work because of social media

riskified.com prevented me from making a purchase at one of their clients' e-stores, one reason being that my e-mail address was unknown, and another because they were unable to find an on-line presence

**Comment by Thorin-Oakenpants** [comment]
[Nosedive](https://en.wikipedia.org/wiki/Nosedive_(Black_Mirror))

**Comment by StanGets** [comment]
Media.com's article, _Privacy Inequality: The Most Brutal Form of Inequality You’ve Ever Imagined_
If I happen to maintain my position among _The Protected_, if I happen to not be already amid _The Predictables_ it's because I'm aware and behave accordingly, those two conditions are required.

Being aware: is this a grace, a faculty of intelligence, a product of one cultural environment?
Acting accordingly: this is obviously the fact of will, the refusal of defeatism.

Many ignore the value of their privacy and how it's being invaded, others know but don't care, others care but are defeatist.

Several parameters for several scenarios. Privacy is certainly not a cultural phenomenon, I believe it is a natural component of ethics and psychology essential requirements. But at the same time some consider their private life in proportion of cultural standards which leads to, i.e. people talking about their very private life on talk-shows for instance, as if they considered that if others do it, why not them: the sheep syndrome which is a high-risk factor: manipulation is also getting people to believe that everyone else except them does it, so why would they be an exception?

Let us think by ourselves, to start with.

**Comment by KOLANICH** [comment]
>Social Credit System

Please stop calling it this way. This is not a kind of credit a moneylender gives to people who ask for it. It is a kind of credit a robber in a dark alley gives to its victim when requesting to go to the flat to take the money stored there too, and this kind of "credit" is usually called not credit, but robbery and blackmail. Black must be called black.

**Comment by atomGit** [comment]
> > Social Credit System
> 
> Please stop calling it this way.

i didn't - the Chinese government did - i think you misunderstood what i wrote

i'm quite aware of the surveillance taking place in so-called "developed" countries and how it affects society

**Comment by KOLANICH** [comment]
>i didn't - the Chinese government did

It doesn't matter who have started to call it this way first. What matters is that, it is incorrect to call it this way, even if you are only repeating the said by someone.

**Comment by Thorin-Oakenpants** [comment]
OMG ... stop calling it Facebook .. its not a book :trollface: :)

If you care to read all 21 (they're short) .. Rick Falkvinge's posts re-enforce what was said about kids growing up with no sense of what privacy is, and surveillance is the norm. FFS, people fought and died for privacy.

- First one: https://falkvinge.net/2017/12/16/analog-equivalent-privacy-rights-01/
- But they're not linked, so go to the privacy tag category https://falkvinge.net/topic/principles/civil-liberties/privacy/

And where was Winston .. they couldn't see him... but they knew where he was .. he was in the *one* place they couldn't see him .. so they knew exactly where he was ... and no, I have never read 1984

**Comment by claustromaniac** [comment]
> I have one or two addons in mind, but really don't have a heart to ask, since I know you are busy with other stuff.

I can't promise I will be able or willing to make them, but I don't mind you asking :)

**Comment by crssi** [comment]
@StanGets thank you for pointing this out

> Coloring the urlbar background given the site's security status is easily performed right from the user's userChrome.css file thanks to a few lines, available i.e. with Color your URL bar based on website security

I have played with it after your post and if you interested, here is the result [https://github.com/crssi/Firefox/blob/master/userChrome.css](https://github.com/crssi/Firefox/blob/master/userChrome.css)

Cheers

**Comment by crssi** [comment]
@claustromaniac 

Thank you. <3

I was thinking about a "fork" of TC, which would be less complicated (if possible even not configurable) with some functions stripped off and some added. But I know this would be a complicated project.
If you are interested, I will dump more details.

Cheers

**Comment by StanGets** [comment]
@crssi ,

> I have played with it after your post and if you interested, here is the result https://github.com/crssi/Firefox/blob/master/userChrome.css

Nice work. I copy, report while others enhance... c'est la vie (my universal answer for laziness! .. but mainly for a great of knowledge in the coding area!)

**Comment by atomGit** [comment]
> I was thinking about a "fork" of TC ...

interesting - because you're thinking about this, i assume the API isn't going to see the addition of the functionality necessary to clear all storage per-domain any time soon?

**Comment by crssi** [comment]
Didn't assume anything, but until TC works I really don't care about those APIs, since all the browsing done here is in a temporary containers, so all the storage's are destroyed sooner or later.

**Comment by atomGit** [comment]
not sure you understood what i wanted to say ... what i meant was, if the API is expanded in the near future to include removal of all storage per-domain, then wouldn't FPI be enough without containers?

if not though, then i'm certainly interested in your container idea

**Comment by crssi** [comment]
^^You already follow the topic https://github.com/ghacksuserjs/ghacks-user.js/issues/395#issuecomment-453128207.
I learned there that FPI still breaks a cat shit (cross domain logins..., but I haven't observed any... yet).
So, with a "correcly" configured TC (and usage), we can still get similar "protection" even with FPI disabled and without breaking cross domain logins.

**Comment by atomGit** [comment]
@crssi @claustromaniac 

re: coloring identity box/url bar based on status

the 'Librefox HTTP Watcher' add-on didn't work for me - it colors both the search and url bar red and leaves them that way for ever more after, for instance, you load an about: page

if you guys have this problem, or are looking for an alt solution, see [CustomCSSforFx](https://github.com/Aris-t2/CustomCSSforFx) and look in /classic/css/locationbar/identitybox_colors.css - he's coloring the ID box (only) for every available status apparently

**Comment by crssi** [comment]
@atomGit  I am using now my custom css, see https://github.com/ghacksuserjs/ghacks-user.js/issues/492#issuecomment-453474612
But upper link is a great resource. :+1: 
Thank you

**Comment by StanGets** [comment]
@atomGit thanks for spotting that specific css. Aris-t2's /CustomCSSforFx is a haven of most valuable CSSs for tweaking Firefox's appearance, either used such as either, as I do, selecting here and there a given css we'll then paste into our userChrome.css. Let's not forget that is is Aris-t2 who had developed the famous _Classic Theme Restorer_ dedicated to restoring the old look to the _Australis_ Firefox back when legacy add-ons were still the rule ...

**Comment by atomGit** [comment]
indeed - CTR was something i could not and would not live without - `https://github.com/overdodactyl/ShadowFox` is another nice theme - this is a dark theme that this takes much of the bite out of the 'white flash' syndrome when you load a new tab

**Comment by claustromaniac** [comment]
@crssi,
> I was thinking about a "fork" of TC, which would be less complicated (if possible even not configurable) with some functions stripped off and some added. But I know this would be a complicated project.
> If you are interested, I will dump more details.

Actually, there are some things about TC that I would like to change myself, and I don't use like half of its features or more. Realistically, I don't have the time for a project of such complexity for now, though. :crying_cat_face:

@atomGit,
> @crssi @claustromaniac 
> 
> re: coloring identity box/url bar based on status

You didn't mean to @ me there, right?

**Comment by crssi** [comment]
@claustromaniac 

> Actually, there are some things about TC that I would like to change myself, and I don't use like half of its features or more. Realistically, I don't have the time for a project of such complexity for now, though.

Understandable, when (if) you will have a time and a wish, we can do a little brainstorm about features. :cat: 

Cheers

**Comment by Atavic** [comment]
https://github.com/april/certainly-something is an extension to view information about the current state of your HTTPS connection, using the TLS Info API implemented from Firefox 62.

Maybe it's an active alternative to the legacy extension *SSleuth*?

**Comment by Thorin-Oakenpants** [comment]
FYI: https://www.zdnet.com/article/websites-can-steal-browser-data-via-extensions-apis/

**Comment by StanGets** [comment]
FYI I've just added an issue on Mozilla Addons' GitHub repository concerning Webextensions in general :
[Webextensions don't respect cookie policy](https://github.com/mozilla/addons/issues/895)

i'd appreciate your comments on this.

**Comment by Thorin-Oakenpants** [comment]
^^ keep me apprised of any bugzilla that comes from this

**Comment by StanGets** [comment]
Concerning bugzilla I'd have to open an account, not that I can't but i admit that I always considered bugzilla as reserved for pros: just reading some reports there makes me believe a strong technical background is required to validate one's descriptions, arguments and so on. But I'll think about it. I believed that Mozilla Addons' GitHub repository would handle the issue...

> the request doesn't have a "tab" with a domain in it 

This gives me a path to start conceptualizing the reasons of the mighty power dedicated to Webextensions but IMO whatever privileges they shouldn't prevail on Firefox's default settings such as blocking cookies when blocking cookies is the general cookie behavior a user has decided.

If I get this issue on bugzilla I'll of course report here the number. I admit being aware (I mean; frightened, let's admit it) of a lasting trial-like debate on Bugzilla where my lack of technical skills could possibly end up by a refusal of Mozilla devs to consider my request on the ground of explanations I wouldn't understand: I think I need a good lawyer!

Thanks @Thorin-Oakenpants

**Comment by Thorin-Oakenpants** [comment]
bugzilla for pros only .. not really.  The only rule for filing bugs in my experience (been around for almost a 100 years) is to always provide STR (steps to reproduce) - then it is up to triage in this case.

What you said in the linked mozilla addons repo is fine - has STR, expected results, actual results. Just copy/paste that in with an app title. I think this is a very important bug to file. And I don;t expect you to be able to find an existing bug - I don't even think I could.

**Comment by StanGets** [comment]
OK, I've filled the bug at [1525917 - Webextensions don't respect cookie policy](https://bugzilla.mozilla.org/show_bug.cgi?id=1525917)

We'll see, first if it's accepted, secondly the answers i'm anxious to discover.

Thanks @Thorin-Oakenpants for providing that push I needed :=)

**Comment by Thorin-Oakenpants** [comment]
add another comment and link back to the issue at the repo you opened

**Comment by Atavic** [comment]
Extensions that use the [Cookies API](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/cookies) declare this behaviour in the [manifest file](https://github.com/gorhill/uBlock/blob/master/platform/firefox/manifest.json#L80).

^*uBo* uses uses ```<all_urls>``` [Permission](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Work_with_the_Cookies_API) to access any protocol and domain.

**Comment by Thorin-Oakenpants** [comment]
Hah .. they bumped it to P1 - maybe a decision is pending soon

The problem I have is that it allows third party cookies, which are not required for an extension to have it's own localStorage, IDB etc. I have no idea if they will WONTFIX this
- I don't know if they can even distinguish between webextension calls versus other ones (like mozilla's internal ones vs web pages etc)
- I don't know if some extensions use multiple domains to function and this would break them - e.g imagine if easylist.to refused to send a list because it wasn't allowed to set a cookie
- what about extensions that sync data to accounts with logins?

IMO, since they made WE calls privileged content (legacy extensions such as uBO could see these, WE's cannot), then there is no control over rogue extensions doing all sorts of shady things - but they could do that anyway without the need for third party cookies AFAIK (I can think of a dozen evil things)

I'm not too worried about it, because we vet our extensions, and they come from respected devs - that said, I was surprised by this, but I do remember it being mentioned during the buildup to quantum - just a shame there's no way for an extension to monitor other extensions

we're going to need robots to service the robots who service the robots...

**Comment by StanGets** [comment]
Indeed modifications have to be feasible and meet Mozilla's agreement : not sure which of the two is the hardest challenge.

As I see it Firefox's storing is the culprit of all these complications

First, the very concept of allowing sites to store anything else than cookies: why? Don't tell me it's for bandwidth, then why? Is it that some Firefox features require that sites be allowed to store data in the user's profile? localStorage, IndexedDB ... why?!

Secondly, concerning IndexedDB, why is it that Webextensions store their date in the same user's profile sub-folder as sites, `[PROFILE]\storage\default\`?

Why not two different policies and storage folders, i.e.
`[PROFILE]\storage\extensions\`
`[PROFILE]\storage\websites\`

I don't know if it's only me but the candid view of a novice is that the general idea would be that the concept is to ignore the distinction between data originating from the browser and data downloaded from sites, integrating both.

> I don't know if they can even distinguish between webextension calls versus other ones (like mozilla's internal ones vs web pages etc)

Yeah. Isn't this simple question relevant of the chaotic, unclear, messed up implications of the very structure of Firefox storage? The issue I mentioned on BugZilla, naively as a non-techie, stirs the inconsistencies of Firefox's very build, IMO. There's something wrong, maybe not in the code (how should I know?) but in the way things and especially storing have been thought.

Mozilla, as all developers i guess, aims to make complicated things appear simple to the user... but no one up to now has been able to combine what is incompatible and trying to do so leads to problems such as Webextensions laying cookies of sites they call even when the user has chosen his cookie behavior to block them ALL.

Wait and see ...

**Comment by Thorin-Oakenpants** [comment]
> First, the very concept of allowing sites to store anything else than cookies: why? 

Because they have valid use cases. It's part of the standards. But they control it thru the cookie permission - a site doesn't have to actually set or use a cookie, it just needs that permission.

> Why not two different policies and storage folders

Because that would create twice the work. You don't need to separate the location of the data files, you simply tag the data with an appropriate flag. e.g web extension data is already definable by the fact the folders start with `moz-extension`. Same reason we don't have separate files for 1st party cookies vs 3rd party. Or separate files per domain.

> rah rah

I'm not an expert on the internal workings of FF, or even the correct terms to use. The key issue for me is that they can differentiate extension (non web page) calls vs Mozilla internal ones vs actual web pages. I think they call this a `principal` - i.e what is calling the shots?

And this is where it's too complicated for me, I'd have to be Moz dev working in this area to know enough. Just because the principal may be an extension, doesn't mean it isn't an open web-page (but they can probably know it's meant for an open tab/web page). And I don't know what they cal this either, I'll just call it a target.

I look at extensions that use logins: lastpass for example. It's NEEDs 3rd party cookies AFAICT, because, without a webpage, within the extensions options, it can log into your lastpass account online, and get your data for local display and editing etc.

But yes, there are inconsistencies: as clearly seen that extensions don't obey cookie rules. If they did, then I see a lot of breakage, and users would have to open the manifest of each UUID and apply site exceptions.

The cookies/site data user interface is all about web sites. The old UI used to say something like **Accept cookies from websites**. It is not meant to be about extensions. Imagine if a user decides to block all cookies and suddenly extensions break, we'd be back to [this](https://github.com/ghacksuserjs/ghacks-user.js/wiki/4.1.1-Setting-Extension-Permission-Exceptions) which is not a user friendly acceptable solution.

It really comes down to the Moz devs on how they want to classify this. Yes extensions can do web calls, and yes, they need a 1st party cookie to store their options, and probably 3rd party in a lot of cases, in order to work. And the current solution is probably the best - loo, extensions can do lots of things, a cookie is not the worst thing in the world. We have to vet extensions anyway because they are so powerful.

Note to self: I wonder if the cookie & related storage (e.g easylist.to via uBO updating filters) is assigned an Origin Attribute?

**Comment by StanGets** [comment]
The very concept of allowing sites to store anything else than cookies

> Because they have valid use cases. It's part of the standards.

They may valid use cases (I linger to know which ones) but meanwhile allow sites to lay data in my IndexedDB storage folder without the reason being obvious to me.

If I've set the cookie behavior to block all by default with exceptions set by me it is because of what allowing cookies leads to: why does bostonglobe.com lay date in my IndexedDB, why does youtube.com as well? Maybe we're in the same scenario as plain cookies where sites use them even when not required: a valid feature abused by some, many websites.

I just do not want and do not accept a website downloading data in my browser profile without my explicit authorization. Hence blocking all cookies by default but more: should I be interested by a site to the point of accepting its cookies that I'd forget that site should allowing its cookies extend to that site creating its indexedDB in my profile. Same with localStorage: no persistent cookie (Allow exception) for a site that lays data in my localStorage (webappstore.sqlite) which I clean anyway once Firefox is closed).

In other words, yes for basic cookies, no no and systematic no to persistent localStorage (never persistent with session cookies) and ultimately no to indexedDB in my profile for anything else than that of the Webextensions i've installed.

What we all observe is that a browser includes features which are meant for the best user experience (I do not have in mind developers aiming to trick users on the basis of industry lobbying requirements, no conspiracy theory) and that these valuable features are exploited by some websites for their own profit and not for the user's advantage. Hence the amount of defense mechanisms developed by users, by Ghacks-user.js itself to start with, aiming at preserving the best and controlling maybe not the worst natively but the worst as what sites occasionally do with the best.

And this is only the beginning, IMO. Whatever the best intentions the trend is and remains digging into users' life. period.

**Comment by Thorin-Oakenpants** [comment]
Yeah look. They are valid mechanisms, but like everything else, they get abused.

Personally, I have had cookies blocked by default for the last 5+ years. I've allowed sites I log into a cookie (about 10 sites over the years, currently just four), I've allowed about 6 more sites a session cookie to function (and blocked them in uMatrix in headers). That's about it. I've never seen any persistent localStorage or IDB shit in all that time, and the web works for me just fine. Just think of the tens of millions of things I didn't have to clean up or have tracking me

**Comment by StanGets** [comment]
That's what I do as well. I just happen to find this way of proceeding a bit radical.
Anyway, "no woman no cry" and "no browser no cry". We love 'em both.

**Comment by Thorin-Oakenpants** [comment]
@Kraxys said
> Just discovered the FF addon [Site Bleacher](https://addons.mozilla.org/en-US/firefox/addon/site-bleacher/). It automatically clears coolies, local storage **and indexedDB** as soon as the les tab relative to a domain is closed. Far less options than CAD but Site Bleacher clears indexedDB, what CAD can't.

I said
> there is no extension API available to clear IDB by domain
> 
> > and Indexed DBs are removed when first site tab is open, so sites are unable to access data from previous session
> 
> if I read that correctly, it says IDBs (plural) are removed when you visit your first website after opening the browser - see the words "from previous session". I don't think English is the author's first language - that sentence is very unclear.
> 
> You can already do this, btw, by clearing data on close

@crssi said
> I read it as when fist tab of specific domain is opened the IDB is cleared for this very domain, which makes sense

I said
> Why don't you test it

@crssi said
> Well, yes and no. If WE injects a code to delete IDB into the visited site, then it can be deleted.
I assume that the "Site Bleacher" does that and deletes IDB if this is the first tab to open that particular site/domain.
> 
> I have tested the Site Bleacher... not that I need it, since I am using TC for that purpose, but Site Bleacher is successful in deleting IDB.
> 
> Did you test it?

**Comment by Thorin-Oakenpants** [comment]
> Did you test it?

No. I don't have time. I'm not an expert, but I have questions about this
- script injection doesn't sound good (this could expose the UUID? the function can be detected? issues with CSP? issues with rolling the dice with other extensions?)
- it only works on pages that allow JS (but then JS would be needed to read existing IDB entries?)
- timing, overhead - it has to inject a script and run it before the site run's it's script
- how does it detect the domain IDB entries - I am not an expert on IDB, but it is NoSQL - is there a way to enumerate all databases for the current domain? When I coded some a little while back, i had to provide the data store name

**Comment by StanGets** [comment]
I've installed and use this _Site Bleacher_ Firefox extension in place of previous _ForgetMeNot_.
If _Site Bleacher_ indeed wipes a site's cookies and localStorage (unless whitelisted by _Site Blocker_) it fails to remove any IndexedDB data  set by that site. I've tried with my two testing sites which both lay their IndexedDB when cookies aren't blocked for them, bostonglobe.com and youtube.com and in neither case were their IDB cleaned up, unfortunately.

But _Site Bleacher_ does the job perfectly well for cookies and localStorage, and weighs only around 50KB I think.

**Comment by Thorin-Oakenpants** [comment]
> once the site closed, it fails to remove any IndexedDB data set by that site

It cleans it when you open the site (but not subsequent pages if you already have tab of it open)

**Comment by StanGets** [comment]
> It (supposedly) removes it when you open the site (but not subsequent pages if you already have tab of it open)

If I remember correctly, but I'd have to test again, it may have removed bostonglobe's IDB but I'm sure it didn't remove youtube's IDB. There's also something special with youtube' IDB foler name which ends with something like '3rd-party' ...

I'll test again right now. Stay tuned :=)

**Comment by StanGets** [comment]
My 1st comment confirmed : neither bostonglobe nor youtube have had their IDB wiped by Site Bleacher.

Usually, default, my network.cookie.cookieBehavior = 2 = block all
For the test I've set network.cookie.cookieBehavior = 1 = Block third-party cookies

Opened bostonglobe.com and IDB folder name was:
> https+++www.bostonglobe.com^firstPartyDomain=bostonglobe.com

Opened youtube.com and IDB folder name was:
https+++www.youtube.com^firstPartyDomain=youtube.com

=> Neither wiped by Site Bleacher.

I use in fact Site Bleacher in order to create the 'Allow Temporary Cookie' which was handled by some legacy add-ons, that is accept cookie when on site then remove, which was the 3rd option afer what we know of session cookies and allow cookies (which remain after FF restart).

Global cookie policy : block all
Exceptions:
Allow and keep (3 sites) or Allow for session (several because they require cookie authorization to display correctly (i.e. userstyles.org) either because I want to log in.

For these exceptions I don't want the cookie to remain once i've quit the site => Site Bleacher or ForgetMeNot.
I know these cookies (set with network.cookie.lifetimePolicy = 2 = session) would be removed once FF closed, but I just don't like cookies following me even within the session only when in fact I needed them only for visiting the site.
Mastering the scenarios at 100%

**Comment by crssi** [comment]
@Thorin-Oakenpants 
I have never said that WE injects the code. I have used **IF**. Author should tell us or we should peek into the code to be sure what method is used.

The question was, does this WE clear IDB. And my tests shows that it does.
I didn't comment that the method is good and what implications are behind, since that was not a question at start.

@StanGets The fact that the IDB files are there, does not mean that those IDB's were not data cleared.

TEST 1:
1. Start with plain vanilla.
2. Visit the https://www.hotcleaner.com/cleaning-software-test.html and https://demo.agektmr.com/storage/
3. Make some choices
4. Close both tabs
5. Revisit sites and observe the results

You will see that IDB and LS and Cookies was filled in the previous visit and not cleared.

TEST 2:
1. Start with plain vanilla.
2. Install Site Bleacher WE.
3. Visit the https://www.hotcleaner.com/cleaning-software-test.html and https://demo.agektmr.com/storage/
4. Make some choices
5. Close both tabs
6. Revisit sites and observe the results

You will see that IDB, LS and Cookies were cleared and does not show the data from the last visit on revisit.

**Comment by Thorin-Oakenpants** [comment]
I haven't got time to look at this. This sucks, I **had** a bookmark for testing IDB. You added a title, and then some text. Add as many as you want. When you came back it would display your stored data. This is all you need - because what you entered was unique

https://demo.agektmr.com/storage/ looks like it might do the job for a test.

**Comment by crssi** [comment]
I have looked into a source code and I guess my guessing was right... the code injection is involved.

This is the code injected:
``` js
    const oldIndexedDBOpen = window.indexedDB.open;
    function newIndexedDBOpen(arg1) {
        const e = new CustomEvent("new_indexdb", {
            detail: arg1
        });
        document.dispatchEvent(e);
        return oldIndexedDBOpen.apply(this, [arg1]);
    }
    newIndexedDBOpen.bind(window.indexedDB);
    window.indexedDB.open = newIndexedDBOpen;
```

This opens all your valid questions about if this is good or not.
As I said before, I am not using this extension and I am not intent to use it, since everything is covered by TC here.

Cheers

**Comment by Thorin-Oakenpants** [comment]
This is it: https://static.raymondcamden.com/demos/2014/feb/7/index.html#/home

Now use that to stick in unique data

**Comment by crssi** [comment]
^^ I have now tested on the page you suggested and revisiting the page comes out clean, IDB cleared.

But since we have a working solution already, I would not analyze this extension anymore... and it raises too many other valid questions... detection of this extension, CSP, does it work when JS is disabled... and on and on.

**Comment by Thorin-Oakenpants** [comment]
First of all, it doesn't clean up after itself, because naturally, it only cleans when you open the domain, and it can't wipe folders etc. But you can do all that, sanitizing, on close. There is an option to whitelist until close (in other words, session a domain), but to me that's the same as whitelisting, because you clear on close.

The very first time I tested it, the extension didn't do anything, until I clicked on settings on the icon. And even though I didn't change anything, THEN it worked. I hope that was just a one off and not for every domain.

```
- make sure nothing is deleted on close (keep all cookies and shit)
- visit test page
- add a note, e.g Chocolate Cake
- close page
- open page (see your note is still there)
- close browser
- open browser & visit page (see your note is still there)
- close page (don't leave it open)
- enable/install extension
- close browser (important)
- open browser, go to test page - your data is still there
   ^^ FAILED on first attempt
- after that it seemed to work as expected
```

Here are my french fries
![idb](https://user-images.githubusercontent.com/16656956/53335578-c5e77700-38f3-11e9-90c9-d39dca4e6663.png)

As long as I closed all instances of the test page, the next time I opened it, it was emptied. This will leave persistent disk data (forensics), but will prevent the website from accessing it next time you visit.

I do not know about the rest of the potential issues such as detecting script injection, function names, leaking UUID, and causing issues with other extensions, or even failing on sites with CSP or whatever.

That said, it's a metric fuck-tonne better than C-AD regards leaving orphaned data around to be re-used to track you

**Comment by Thorin-Oakenpants** [comment]
> does it work when JS is disabled

Well, without JS, the data can't be read either

**Comment by Thorin-Oakenpants** [comment]
I do have questions over the third party data, it should be cleaning ALL data by first party IMO, and I'm not in a position to test it all - but using the storage inspector and a site that uses IDB on 3rd party should would be a start

**Comment by crssi** [comment]
First time it failed because you had the test page opened in time you had installed WE. ;)

> Well, without JS, the data can't be read either

Sure it is, but what when you enable JS in the middle of loading page?

**Comment by Thorin-Oakenpants** [comment]
Bollocks. The page was not open in the session at all, and the extension was up and running and installed. Repeat from what I said earlier

```
- close page (don't leave it open)
- enable/install extension
- close browser (important)
- open browser, go to test page - your data is still there
   ^^ FAILED on first attempt
```

pages closed before installing extension and not opened until after extension on
closed the browser even
opened the browser, first new session, extension is on .. and it failed, but only that first time. The button had a red [1] on it and everything

---

> but what when you enable JS in the middle of loading page

in the *middle of loading a page*? wot? That's not normal bro, see a doctor.

**Comment by atomGit** [comment]
> There is an option to whitelist until close (in other words, session a domain), but to me that's the same as whitelisting, because you clear on close.

you may be referring to the same issue i raised with the dev - this is his reply...

> Unfortunate wording, changed to "whitelist tab" only, means cookies won't be deleted on domain changes like usual, helps with sites where you are redirected to another domain for login

**Comment by crssi** [comment]
> in the middle of loading a page?

I was thinking not to underestimate a standard user diversity of usage.

> That's not normal bro, see a doctor.

With all the respect, lady, from the lately insulting and aggressiveness I am more thinking to see out of here. Obviously I am just wasting your time and I am sorry for that.

Cheers

**Comment by atomGit** [comment]
that comment was rather rude pants - i know you mean well, but you can be quite abrasive at times - and i know how to spot those people because i'm one of them unfortunately

**Comment by Thorin-Oakenpants** [comment]
You're just reading too much into my words. I'm not being aggressive or insulting at all - at least not the way I said it in my head as I typed it.

Yes you're right to always think of possible situations - the permutations, etc. But loading JS partway thru a page load seems almost impossible. In most cases a user would enable js and reload - e.g changes to tweaking uM, uBO, etc. I'm not even sure *how* you could enable it part way thru a page load

**Comment by Thorin-Oakenpants** [comment]
> Sure it is, but what when you enable JS in the middle of loading page?

Here you go. If you have JS disabled and load the page. Then you enable JS and reload the page, does the extension then treat the reload as the FIRST time it was opened? Probably not.

**Comment by Thorin-Oakenpants** [comment]
Here's the thing guys and gals, and I'll try and be more selective with words in future

Slang/idioms, tone, and emotion, do not translate well across text. To me what I said was something lighthearted and slightly comedic - in my realm of real life people, it's not offensive (seeing a doctor for a non medical issue is part of the joke) - it's just a way of saying, I don't think you thought that thru, would you care to explain it further.

I am also incredibly extremely busy, especially right now, but mainly over the last 4 months (which is not anyone's fault, just laying some background), and part of my work methodology is practicing a zero inbox. So when I get notifications, I like to deal with them and delete them. I also use the open issues in this repo as my ToDo list or reminders. When in the middle of a lot of work, I can often make quick replies, and throw the ball back in the other person's court - hence why I kept telling you guys to test it (rather than me doing it).

Been here for two years. Everyone should know that I never have any problems with anyone (except that one guy, you know who), and I often go the extra mile in working things out, or explaining things. Earthlng even recently told me off for trying to do to much.

So that said, it's up to you, individually, as to whether or not, you want to be part of this repo. Earthlng and I have had arguments in the past, and we get tetchy or frustrated with each other, but we know there's nothing personal in it. And likewise, I know everyone here is asking genuine questions, or contributing - that's what collaboration is. There are no stupid questions or statements, at worst, just misunderstanding. And no-one is right all of the time (sometimes I forget that bit).

just going to @crssi here in the hope he will see it. Apologies dude. You are welcome here, you have collaborator status, and you've been here from the start.

**Comment by crssi** [comment]
Noted. Next time I will tell you "STFU, grab a beer, get naked and run around appartment singing some songs. :smile:" (<- in a good sense) :heart:

**Comment by atomGit** [comment]
@Kraxys - re: [Change Timezone](https://addons.mozilla.org/en-GB/firefox/addon/change-timezone-time-shift)

i use a vpn and when i test at [browserspy.dk](http://browserspy.dk/date.php) it's detecting my TZ as GMT, which is not my real TZ, and this is without the add-on - why this is the case i don't know off-hand

be curious what your results are if you also use a vpn
